src/OrchardCore.Modules/OrchardCore.Media/SecureMediaPermissions.cs

```diff
@@ -0,0 +1,155 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Options;
+using OrchardCore.Media.Services;
+using OrchardCore.Security.Permissions;
+
+namespace OrchardCore.Media
+{
+    public class SecureMediaPermissions : IPermissionProvider
+    {
+        // Note: The ManageMediaFolder permission grants all access, so viewing must be implied by it too.
+        public static readonly Permission ViewMedia = new("ViewMediaContent", "View media content in all folders", new[] { Permissions.ManageMediaFolder });
+        public static readonly Permission ViewRootMedia = new("ViewRootMediaContent", "View media content in the root folder", new[] { ViewMedia });
+        public static readonly Permission ViewOthersMedia = new("ViewOthersMediaContent", "View others media content", new[] { Permissions.ManageMediaFolder });
+        public static readonly Permission ViewOwnMedia = new("ViewOwnMediaContent", "View own media content", new[] { ViewOthersMedia });
+
+        private static readonly Permission s_viewMediaTemplate = new("ViewMediaContent_{0}", "View media content in {0}", new[] { ViewMedia });
+
+        private static Dictionary<ValueTuple<string, string>, Permission> s_permissionsByFolder = new();
+        private static readonly char[] s_trimSecurePathChars = ['/', '\\', ' '];
+        public static readonly ReadOnlyDictionary<string, Permission> PermissionTemplates = new(new Dictionary<string, Permission>()
+        {
+            { ViewMedia.Name, s_viewMediaTemplate },
+        });
+
+        private readonly MediaOptions _mediaOptions;
+        private readonly AttachedMediaFieldFileService _attachedMediaFieldFileService;
+        private readonly IMediaFileStore _fileStore;
+
+        public SecureMediaPermissions(
+            IOptions<MediaOptions> options,
+            IMediaFileStore fileStore,
+            AttachedMediaFieldFileService attachedMediaFieldFileService)
+        {
+            _mediaOptions = options.Value;
+            _fileStore = fileStore;
+            _attachedMediaFieldFileService = attachedMediaFieldFileService;
+        }
+
+        public async Task<IEnumerable<Permission>> GetPermissionsAsync()
```

Comment:
I'm open to suggestions... 😇

Reply:
I don't have a better idea than caching them, unfortunately.

Some suggestions:

- The hard part will be invalidation, but that we can start with a simple "clear the whole cache if there are any folder changes".
- `IDistributedCache` can be used to store the cache.
- `IDistributedLock` might be needed to synchronize rebuilding the cache in a multi-node setup.

Reply:
Hmm, caching here will introduce some complexity, even if the secure media feature is not used. Maybe we should just live with this performance hit, as it's in the admin only.

We would at least need to add some signaling when folders are changed in the admin. As you said, it must be distributed too. Maybe we can use an `ISignal` like in background tasks in the media file stores create- and delete folder methods. Those signals can also be just raised by the secure media feature, but in that case we need more media events (there is none for creating folders).

Reply:
Storage I/O can be the narrowest bottleneck for a web app, e.g. if you host it on Azure App Services and/or use Azure Blob Storage for Media storage (see https://github.com/OrchardCMS/OrchardCore/issues/14859 and https://github.com/OrchardCMS/OrchardCore/issues/15016). It's not necessarily about increasing the latency of the request (though with a Blob Storage request easily taking 10-20 ms, that would still increase an admin request's latency by 5-20%, which is quite significant), but consuming from the very low IOPS limits shared by the web app's binaries, Media and other data files, and the OS too.

You don't necessarily need signals. That's required if you want kind of an event bus and/or where you read/write the cache and where you invalidate the it are different and independent projects. In this case though, everything would be in the Media module, and potentially even within the same service.

Reply:
> Storage I/O can be the narrowest bottleneck for a web app, e.g. if you host it on Azure App Services and/or use Azure Blob Storage for Media storage (see #14859 and #15016). It's not necessarily about increasing the latency of the request (though with a Blob Storage request easily taking 10-20 ms, that would still increase an admin request's latency by 5-20%, which is quite significant), but consuming from the very low IOPS limits shared by the web app's binaries, Media and other data files, and the OS too.

I'm not questioning that, but only for the admin portal and if you use the admin menu with permissions. 

> You don't necessarily need signals. That's required if you want kind of an event bus and/or where you read/write the cache and where you invalidate the it are different and independent projects. In this case though, everything would be in the Media module, and potentially even within the same service.

I was thinking about caching permissions locally and use signals to invalidate that cache. That was the idea I outlined very roughly above. Alternatively, we can just cache all folders in a distributed cache. Not sure about the way to go, maybe caching the folders is a simpler solution (I don't think adding permissions to the distributed cache directly is a good idea).

If you guys absolutely want to have caching here, I can add that. But it would require more changes to the original media module and it may therefore be better to separate that work into another PR. At least adding more events to the media module should be separated from this PR I think.

Reply:
The admin portal can be a significant part of the requests going to the app, if not even the larger one (in the apps mentioned under https://github.com/OrchardCMS/OrchardCore/pull/15173#discussion_r1552563766, actually the majority of requests go to the admin, since they're management portals, not simple websites). I'm afraid this feature will be prohibitively expensive in such cases.

Note that basically all admin menus in any OC instance will use permissions.

Reply:
Alright, I will come up with something 😊

Reply:
I added some basic caching. It's a rough idea to get some feedback from you, especially I'm not sure about versioning the `IMediaEventHandler`. Thanks for having a look 😊

Reply:
Thank you! I did a quick review (will do that in-depth for the whole delta once you're done).